[RFC] Engineering guiding principles

orta opened the pull request:
This tries to make a spiritual successor to the [Artsy objc.io](https://www.objc.io/issues/22-scale/artsy/) as a part of making more of artsy dev culture explicit instead of implicit.

I took a look at some of our larger tech choices and tried to re-frame some of the terms I've used when speaking publicly to be a consistent set of ideals. I need to do a bit more work on declaring that this is something we strive for, rather than perfect being this.

It's worth noting that this is different from the current org-wide culture survey that @williardx is running.

orta committed:
First shot at engineering principals

orta committed:
More principals

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+# Engineering Principles
+
+We have [company values](what_is_artsy.md), and are working on our shared values with design and product. However,
+when we make decisions on what and how to do engineering we need guiding principals. Things that make Artsy's
```
> make decisions on what and how to do engineering**,** we need guiding principals

Add a comma for clarity?

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+
+### Open Source by Default
+
+We consider the data and relationships that make Artsy difficult to copy, code only represents the past and
```
These are two sentences spliced together with a comma – I think a dash or semicolon would be better.

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+
+We consider the data and relationships that make Artsy difficult to copy, code only represents the past and
+close-to present. A code project should start as open, then move to closed if there are valid business needs that
+need to be hidden.
```
"business needs that need" sound strange, maybe "business needs that require" or similar?

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+close-to present. A code project should start as open, then move to closed if there are valid business needs that
+need to be hidden.
+
+Working in the open allows individual to retain their work and ideas post-Artsy. We can share the exact problem we
```
This should be expanded, as it gets explained in more detail in some of the further reading. Repeating and rephrasing the first sentence could provide clarity and help connect it to the second.

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+Working in the open allows individual to retain their work and ideas post-Artsy. We can share the exact problem we
+see with others, and provide great high ways to teach by direct reference.
+
+Expectations for OSS apps are different than OSS libs, an app has little-to-no desire for community contributions.
```
Another comma splice, maybe replace with a colon?

> different than OSS libs: an app

ashfurrow replied:
I'd also considering italicizing `apps` and `libs`.

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+see with others, and provide great high ways to teach by direct reference.
+
+Expectations for OSS apps are different than OSS libs, an app has little-to-no desire for community contributions.
+Instead it's more of a reference tool. Tools and libraries we can build communities around, but most projects don’t
```
> Instead**,**

Needs a comma.

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+  - http://artsy.github.io/blog/2017/09/04/Introducing-Peril/
+  - http://artsy.github.io/blog/2018/02/02/artsy-apogee/
+
+### Scale your Impact
```
Capitalizing of "Your" should be consistent with other titles.

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+  - Artsy Blog
+  - artsy/meta
+  - Peer Labs
+  - http://artsy.github.io/blog/2015/04/03/artsy-engineering-compensation-framework/
```
We could link to YouTube videos of JavaScriptures or others (eg: https://www.youtube.com/watch?v=yoGielteT64 )

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+  - Learn Swift at Artsy
+  - Artsy Blog
+  - artsy/meta
+  - Peer Labs
```
Link to: https://peerlab.community

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+  - Peer Labs
+  - http://artsy.github.io/blog/2015/04/03/artsy-engineering-compensation-framework/
+
+### De-siloing Engineers
```
This should match tense with other titles, which are imperative. Reword to "De-silo Engineers"?

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+
+### De-siloing Engineers
+
+Provides people the ability to work in many spaces and not feel blocked by another team. Common languages, common
```
Same thing: s/Provides/Provide

ashfurrow commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+### De-siloing Engineers
+
+Provides people the ability to work in many spaces and not feel blocked by another team. Common languages, common
+terminology and common ideas. There's definitely space for experimentation, but within a shared framework. The aim
```
"ideas" -> "idioms" or something a little more substantial? React, for example, is much more than an idea.

ashfurrow reviewed (approved):
Looks great – I would co-sign it all 👍 Just a bunch of nitpicking grammar stuff.

joeyAghion commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+
+We consider the data and relationships that make Artsy difficult to copy, code only represents the past and
+close-to present. A code project should start as open, then move to closed if there are valid business needs that
+need to be hidden.
```
I'm not sure this is a viable business practice. Launching projects as open-source means associating a license with the content and potentially accepting external contributions. Making projects private later is not as simple as checking a box and removing the license file; it's unclear whether we even _could_ make outside contributions private without explicitly re-licensing that content. Plus, any external forks of the previously-public project become orphaned or worse--there becomes a new public ["parent" repo](https://help.github.com/articles/what-happens-to-forks-when-a-repository-is-deleted-or-changes-visibility/#changing-a-public-repository-to-a-private-repository) that we don't control.

orta replied:
> it's unclear whether we even could make outside contributions private without explicitly re-licensing that content

Yeah, in those cases we'd have a few interesting choices:

- Debate moving apps into being open but not strictly OSI-approved OSS ™️  (e.g. a license like MIT, but with a clause saying we can revoke the app at any time)
- Keep MIT, but add CLAs to any OSS repo, it'd bug people the first time they submitted a PR and it'd give us the rights to change license / make private (which is what other companies do)

I guess the framing for the project creation should be useful here, like "start the debate from the perspective of open" rather than "start as open"

alloy replied:
It actually is as simple as ticking a box, because MIT licensed code may be used in closed source projects. The only thing that needs to happen is to not remove the license _or_ annotate the code from external contributors, which could easily be automated as that data exists in SCM anyways.

As for there being a new public parent repo, while that could happen it’s extremely unlikely to ever happen to a repo of ours and additionally isn’t really any different from people having local clones. It’s not something I would really consider in this debate.

joeyAghion commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+
+Take the best dependencies to fit your problem and make them better. Rather than try re-invent the wheel, give time
+to take the best dependencies and make them fit into our use cases. We’re not the biggest team, so the most value
+for someone's time can be achieved by writing the glue code between larger projects.
```
I like the gist of "we're lean, so try to employ the best components from the wider field," but this feels like it's in conflict with also being core contributors to those "best" dependencies. If they're the best, we should contribute but it's unlikely we'll need to "own" them, right?

orta replied:
Great point. 

We use the phrase "own" but it should be more metaphorical I think, for example we're not on the TypeScript core team - but we basically have their digital phone numbers. We probably will never need it, but we can easily make a meeting with them to discuss something critical.

I'll try shift the angle a bit here form a direct "repo access"

joeyAghion commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+  - CocoaPods
+  - The Artsy Omakase
+
+### Incremental Revolution
```
👍

joeyAghion commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+### Minimal Viable Process
+
+As a team grows, so does the need for processes. Find ways to add process without additional work. Build automated
+tools to help others. Use systems like RFCs to broaden the availability of information on process changes.
```
👍

I'd say "Find ways to _get the value_ of process without additional work."

joeyAghion commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,137 @@
+### Scale your Impact
+
+We built an outsized reputation for our size as an engineering team via the above principals. This gives the
+ability to move local impact to larger impact.
```
I really relate to this kind of "everyone taking responsibility for the whole group and product" message. We're a lean team so we must maximize the leverage of our individual efforts by sharing openly, etc. I actually think "leverage" is a nice meaningful term that could be incorporated into the title.

joeyAghion reviewed (commented):
Props for writing these down. Reading and reacting, I wonder if inline comments can lead to "consensus" about such a deep topic, but at least you've started the process.

I selfishly would love to incorporate some of the [Platform team's principles](https://github.com/artsy/potential/blob/master/platform/Principles.md) into this. The most valuable for our work have been:
* "Build for 10x." This helps us decide how much architecture/engineering/planning is too little and how much is too much.
* "Done means done." This makes explicit that we take responsibility for our work product being the _right_ change, performing in production as expected, and being understood and adopted by the necessary stakeholders. I.e., our work is not over when the PR is merged.

orta commented:
WRT consensus, I figured it might make sense to not class this as something mergeable once some commenting has happened, but do feedback etc for a week and iterate - then do an optional for all meeting about wrapping it up and shipping it

orta committed:
Principles Feedback run 1

williardx commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,146 @@
+
+[on route]
+
+### Done Means Done
```
@joeyAghion I like the idea of "Done means done" as described but the name is a little ambiguous to me. When I read "done means done" I associate it with not returning to finished work while the description points more to following through.

joeyAghion replied:
Interesting! I'd heard it this way so many times that the ambiguity disappeared for me, but I see what you're saying now.

I actually disagree with "not returning to finished work" so strongly that I'd like to suggest an additional principle: something like "Expect change." Thoughts?

orta replied:
I feel like "coming back is a new project" could be a sub-part of explaining "done is done"

williardx commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,146 @@
+  - https://code.dblock.org/2015/02/09/becoming-open-source-by-default.html
+  - https://www.objc.io/issues/22-scale/artsy/
+
+### Own Your Dependencies
```
I see two different but related concepts here: "own your dependencies" and "don't reinvent the wheel." You could in theory have one without the other but they go together pretty naturally. Should they be separate, or maybe described with a more general name?

orta replied:
I think own your dependencies at core isn’t quite the same as don’t reinvent the wheel - this speaks to the acceptance that all the code in your stack is something you should contribute to

Don’t reinvent the wheel is a pretty good  way of thinking but it’s basically something most engineering teams should do - not sure if it is something unique enough to us to warrant inclusion

orta replied:
Other than a sentence nod to it, probably

ashfurrow replied:
Yeah, I see "don't reinvent the wheel" as the foundation that the "own your own dependencies" is built upon. A nod sounds like a good idea.

williardx replied:
Agreed that "don't reinvent the wheel" should be baseline but in practice lots of teams end up reinventing the wheel because it's personally more gratifying, or because they believe their needs are unique enough to warrant it. That we take it seriously is kind of interesting!

ashfurrow replied:
Yeah! In contrast to the "not invented here" idea, the objc.io article linked to describes the mobile team's ideology of "proudly discovered elsewhere", which I still admire.

orta replied:
We should re-use that phrase here

orta committed:
Started some work on the 10x and done is done

orta commented:
I've taken a first run "build for 10x" and "done is done"

orta commented:
@sweir27 mentioned "Being Nice is Nice" as one of our principals

ashfurrow commented:
@orta @sweir27 Yeah yeah, I like that a lot. It's kind of... like, unspoken almost? Like, "being nice is nice" is just so much a part of how we work that it's easy to forget it's even there.

alloy commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,159 @@
+Consider ways in which you can expand the scope of impact for your time:
+
+- Avoid private messaging when you could communicate in a public channel
+- Team presentations could be worked for public consumption
```
**Re**worked?

alloy commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,159 @@
+### Done Means Done
+
+Not that when you tell someone it's done that everything is final. More that you can't just throw something over a
+fence, shipping code is one part of an equation but running QA and ensuring stability is just as important. Done
```
The first part of the second sentence should probably be part of the first sentence instead of the second.

sweir27 commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,159 @@
+terminology and common idioms. There's definitely space for experimentation, but within a shared framework. The aim
+is not to stifle creativity, but increase individual scope.
+
+- Examples
```
We could mention the fact that we intentionally try to allow teams/people to not silo from an organizational perspective by allowing ourselves to re-staff/move periodically.

sweir27 commented:
To flesh out the idea I had above (also have to give much credit to my mother who always says `"It's nice to be nice"` which is one of the more helpful life lessons).

---

"Being nice is nice" or "It's nice to be nice" or however we want to say it.

Behind every piece of code is a human, internal or external to Artsy. Give people the benefit of the doubt. Take the time to understand _why_ they made a decision before making assumptions. Realize that you may come across differently on slack/github than in-person and consider how the person on the other side might respond to what you're saying. There's _always_ a nice way to handle a situation, and we strive for that.

orta commented:
Being nice is nice is good for me, I gave a talk on this [back in 2015](https://academy.realm.io/posts/altconf-orta-therox-being-nice-in-open-source/)

orta committed:
More feedback, adds real links

joeyAghion commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,189 @@
+
+A system won't live forever, but you should make decisions with a aim to handle 10x the traffic, number of
+contributors and company revenue. Basically a single order of magnitude of growth everywhere. This can help guide
+deciding on technological choices and figuring the trade-offs between feature work and infrastructure needs.
```
My attempt at re-wording this (take what you like):

> Technology choices should strive to be "optimal" while avoiding over-engineering. When designing systems or evaluating scalability and performance, we aim for today's decisions to withstand 10x the traffic, data, or scale. This realistic horizon helps us balance the need to move quickly with the sometimes-competing need to invest in infrastructure and architecture. It also recognizes that solutions are expected to evolve and be replaced.

orta replied:
I love it, replaced.

orta committed:
More feedback

joeyAghion commented on meta/engineering_principles.md:
```diff
@@ -0,0 +1,189 @@
+
+You can't just throw your code over a fence, shipping code is one part of an equation but running QA and ensuring
+stability is just as important. Done being done means feeling confident in that you've protected your changes with
+tests, ensured deployment works and feel confident in your tools for measuring.
```
My attempt:

> We take responsibility for not just implementing the defined work but ensuring it's the _right_ work, delivered all the way to [and understood by] the end consumer, instrumented, tested, and confirmed to behave as expected in production. This means we document, deploy, and monitor our changes. It's "done" when we can comfortably put it out of our minds and move on to new work.

joeyAghion commented:
I like "being nice is nice" and think that some other "obvious" items might be worth making explicit for new team members:
* shipping early and often
* making the "hard" things easy with CI/CD/devops
* anybody can contribute to anything and our open-source-style workflow

starsirius commented:
This is great! I like that we have examples and further readings for each section! Minor suggestion that it will help me read with a ToC type of summary on the top, given the # of items we have.

While reading through it, it reminds me of the triangle (quantity-quality-time). Maybe that's worth mentioning, too.

mennenia commented:
❤️

joeyAghion commented:
Another principle (from Platform) that felt relevant today as I was reviewing the last and next quarters: "Order matters." I.e., we think carefully about the right order in which to take on big projects. Initial steps can de-risk projects, expose mistaken assumptions, surface feedback, and create opportunities for reuse or other compounding benefits.

orta committed:
Handles more feedback

orta commented:
OK, I think I've addressed all the feedback in here 

WRT the TOC, I'll fix and build something more systematic for the entire upcoming README repo rather than a one-off for this file.

orta commented:
Will merge this on Tuesday if there's no breaking feedback, it's obviously a living document - I'd love some blog posts or links around the last two values so they feel fleshed out

orta commented:
Alright, merging this. 🎉
